p2p/dial.go

```diff
@@ -0,0 +1,207 @@
+package p2p
+
+import (
+	"container/heap"
+	"crypto/rand"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/ethereum/go-ethereum/logger"
+	"github.com/ethereum/go-ethereum/logger/glog"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+)
+
+const (
+	// this is the amount of time spent waiting in between
+	// redialing a certain node.
+	dialHistoryExpiration = 30 * time.Second
+)
+
+// dialstate schedules dials and discovery lookups.
+type dialstate struct {
+	maxDynDials int
+
+	lookupRunning bool
+	dialing       map[discover.NodeID]connFlag
+	lookupBuf     []*discover.Node // current discovery lookup results
+	static        map[discover.NodeID]*discover.Node
+	hist          *dialHistory
+}
+
+// the dial history remembers recent dials.
+type dialHistory []pastDial
+
+// pastDial is an entry in the dial history.
+type pastDial struct {
+	id  discover.NodeID
+	exp time.Time
+}
+
+type dialTask struct {
+	flags connFlag
+	dest  *discover.Node
+}
+
+type discoverTask struct{ results []*discover.Node }
+
+type waitExpireTask struct{ time.Duration }
+
+func newDialState(static []*discover.Node, maxdyn int) *dialstate {
+	s := &dialstate{
+		maxDynDials: maxdyn,
+		static:      make(map[discover.NodeID]*discover.Node),
+		dialing:     make(map[discover.NodeID]connFlag),
+		hist:        new(dialHistory),
+	}
+	for _, n := range static {
+		s.static[n.ID] = n
+	}
+	return s
+}
+
+func (s *dialstate) addStatic(n *discover.Node) {
+	s.static[n.ID] = n
+}
+
+func (s *dialstate) newTasks(running []task, peers map[discover.NodeID]*Peer, now time.Time) []task {
+	var newtasks []task
+	addDial := func(flag connFlag, n *discover.Node) bool {
+		_, dialing := s.dialing[n.ID]
+		if dialing || peers[n.ID] != nil || s.hist.contains(n.ID) {
+			return false
+		}
+		s.dialing[n.ID] = flag
+		newtasks = append(newtasks, &dialTask{flags: flag, dest: n})
+		return true
+	}
+
+	// Compute number of dynamic dials necessary at this point.
+	needDynDials := s.maxDynDials
+	for _, p := range peers {
+		if p.rw.is(dynDialedConn) {
+			needDynDials--
+		}
+	}
+	for _, flag := range s.dialing {
+		if flag&dynDialedConn != 0 {
+			needDynDials--
+		}
+	}
+
+	// Expire the dial history on every invocation.
+	s.hist.expire(now)
+
+	// Create dials for static nodes if they are not connected.
+	for _, n := range s.static {
+		addDial(staticDialedConn, n)
+	}
+	// Create dynamic dials, removing tried items from the lookup
+	// result buffer.
+	i := 0
+	for ; i < len(s.lookupBuf) && needDynDials > 0; i++ {
+		if addDial(dynDialedConn, s.lookupBuf[i]) {
+			needDynDials--
+		}
+	}
+	s.lookupBuf = s.lookupBuf[:copy(s.lookupBuf, s.lookupBuf[i:])]
+
+	// Launch a discovery lookup if more candidates are needed.
+	if len(s.lookupBuf) < needDynDials && !s.lookupRunning {
```

Comment:
Don't we want to allow equality here too? Currently there are two possibilities to get here: either the lookup buffer is depleted or the needed dials.

 * if the lookup is empty, but we need to dial, it will fetch new peers (ok)
 * if the needed dials is 0 but we still have peers is the buffer it will not fetch new peers (ok)
 * if both are empty, it will *not* fetch new peers. IMHO it should, to pre-fetch before there's actually need for them.